src/main.rs

```diff
@@ -201,6 +232,112 @@ fn main() -> ! {
     }
 }
 
+fn check_and_process_control_can_frames(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+    debug_console: &mut DebugConsole,
+) -> Result<(), OxccError> {
+    // poll both control CAN FIFOs
+    for fifo in &[RxFifo::Fifo0, RxFifo::Fifo1] {
+        match can_gateway.control_can().receive(fifo) {
+            Ok(rx_frame) => {
+                modules.brake.process_rx_frame(&rx_frame, debug_console)?;
+                modules
+                    .throttle
+                    .process_rx_frame(&rx_frame, debug_console)?;
+                modules
+                    .steering
+                    .process_rx_frame(&rx_frame, debug_console)?;
+            }
+            Err(e) => {
+                // report all but BufferExhausted (no data)
+                if e != CanError::BufferExhausted {
+                    return Err(OxccError::from(e));
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn check_for_faults(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+    timer_ms: &MsTimer,
+    debug_console: &mut DebugConsole,
+) -> Result<(), OxccError> {
+    let maybe_fault = modules.brake.check_for_faults(timer_ms, debug_console)?;
+    if let Some(brake_fault) = maybe_fault {
+        can_gateway.publish_fault_report(brake_fault)?;
+    }
+
+    let maybe_fault = modules.throttle.check_for_faults(timer_ms, debug_console)?;
+    if let Some(throttle_fault) = maybe_fault {
+        can_gateway.publish_fault_report(throttle_fault)?;
+    }
+
+    let maybe_fault = modules.steering.check_for_faults(timer_ms, debug_console)?;
+    if let Some(steering_fault) = maybe_fault {
+        can_gateway.publish_fault_report(steering_fault)?;
+    }
+
+    Ok(())
+}
+
+fn publish_reports(
+    modules: &mut ControlModules,
+    can_gateway: &mut CanGatewayModule,
+) -> Result<(), OxccError> {
+    // attempt to publish them all, only report the first to fail
+    let mut result = Ok(());
```

Comment:
Contrary to what the comment says, this pattern will only result in the *last* failure being reported.